ILM Make the `check-rollover-ready` step retryable

andreidan opened the pull request:
This adds the infrastructure to be able to retry the execution of retryable
steps and makes the`check-rollover-ready` retryable as an initial step to 
make the `rollover` action more resilient to transient errors.

This is the initial effort to tackle #44135

andreidan committed:
ILM open/close steps are noop if idx is open/close

The open and close follower steps didn't check if the index is closed, open
respectively, before executing the open/close request.
This changes the steps to check the index state and only perform the
open/close operation if the index is not already open/closed.

andreidan committed:
ILM Make the `check-rollover-ready` step retryable

This adds the infrastructure to be able to retry the execution of retryable
steps up to a configurable number of times (controlled via the setting
`index.lifecycle.max_failed_step_retries_count`) and makes the
`check-rollover-ready` retryable as an initial step to make the `rollover`
action more resilient to transitive errors.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/ILM+SLM)

andreidan committed:
Drop unused imports

andreidan commented:
Created this as a draft to get some feedback on the progress. 

What's left is adding the retry count and the flag to signal the transitive error to the `ilm/explain` api and documentation (either as part of this PR or as a followup).

andreidan committed:
ILM add step retries information to explain api

jasontedor commented:
> controlled via the setting `index.lifecycle.max_failed_step_retries_count`

What is the reasoning behind adding a setting for this? If we do add this, why is the default finite ([currently in the implementation](https://github.com/elastic/elasticsearch/pull/48256/files#diff-2741416bb52c2a184979d7a7a94dd753R39-R40), it's 15).

andreidan commented:
> > controlled via the setting `index.lifecycle.max_failed_step_retries_count`
> 
> What is the reasoning behind adding a setting for this? If we do add this, why is the default finite ([currently in the implementation](https://github.com/elastic/elasticsearch/pull/48256/files#diff-2741416bb52c2a184979d7a7a94dd753R39-R40), it's 15).

@jasontedor we have integrations tests that expect the `hot` phase to end up in the `error` state. With the rollover action being retryable we'll flip back the lifecycle execution step to the failed step when an error occurs and the tests might not catch the error state. Controlling the number of retries is mostly for this particular use case where we'll allow 1-2 retries in tests before the `error` state is reached. I don't think there's any production use case where a user would want to modify this setting. 

Regarding being finite or not, I don't have a strong opinion on this, just genuinely scared of unbounded loops. Although the retries are reacting to cluster state changes so there is a bit of backoff involved there (with the future addition of exponential backoff I think it will make more sense to be finite)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CloseFollowerIndexStep.java:
```diff
@@ -32,13 +32,17 @@ void performDuringNoSnapshot(IndexMetaData indexMetaData, ClusterState currentCl
-            },
-            listener::onFailure)
-        );
+        if (indexMetaData.getState() == IndexMetaData.State.OPEN) {
```
Can you explain this check? Is this a separate issue you found?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponse.java:
```diff
@@ -35,6 +35,8 @@
     private static final ParseField ACTION_FIELD = new ParseField("action");
     private static final ParseField STEP_FIELD = new ParseField("step");
     private static final ParseField FAILED_STEP_FIELD = new ParseField("failed_step");
+    private static final ParseField IS_TRANSITIVE_ERROR_FIELD = new ParseField("is_transitive_error");
```
I'm not sure about this name, I think a better name might be `error_auto_retryable` or something similar about how the error *will* be retried.

danhermann replied:
Transient?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponse.java:
```diff
@@ -162,13 +174,17 @@ public IndexLifecycleExplainResponse(StreamInput in) throws IOException {
             stepInfo = in.readOptionalBytesReference();
             phaseExecutionInfo = in.readOptionalWriteable(PhaseExecutionInfo::new);
+            isTransitiveError = in.readOptionalBoolean();
+            failedStepRetryCount = in.readOptionalVInt();
```
These will need versioning protection when backporting.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponse.java:
```diff
@@ -193,6 +209,8 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeOptionalBytesReference(stepInfo);
             out.writeOptionalWriteable(phaseExecutionInfo);
+            out.writeOptionalBoolean(isTransitiveError);
+            out.writeOptionalVInt(failedStepRetryCount);
```
Same comment here about version checks when backporting

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponse.java:
```diff
@@ -288,6 +314,12 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            if (isTransitiveError != null) {
+                builder.field(IS_TRANSITIVE_ERROR_FIELD.getPreferredName(), isTransitiveError);
+            }
+            if(failedStepRetryCount != null) {
```
Super minor nit:

```suggestion
            if (failedStepRetryCount != null) {
```

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/OpenFollowerIndexStep.java:
```diff
@@ -23,13 +23,17 @@
-            },
-            listener::onFailure
-        ));
+        if (indexMetaData.getState() == IndexMetaData.State.CLOSE) {
```
Is this addressing a separate bug?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/Step.java:
```diff
@@ -38,6 +38,13 @@ public StepKey getNextStepKey() {
     }
 
+    /**
+     * Indicates if the step is retryable when it encounters an execution error.
```
```suggestion
     * Indicates if the step can be automatically retried when it encounters an execution error.
```

We should be careful in our wording since any step can be retried with the `/_ilm/retry` API, but only certain ones would be (currently) able to be *automatically* retried.

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/CloseFollowerIndexStepTests.java:
```diff
@@ -110,6 +110,30 @@ public void onFailure(Exception e) {
         Mockito.verifyNoMoreInteractions(indicesClient);
     }
 
+    public void testCloseFollowerIndexIsNoopForAlreadyClosedIndex() {
```
This seems related to the open/close guards above, but I think that should be a separate change?

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponseTests.java:
```diff
@@ -162,11 +168,15 @@ protected IndexLifecycleExplainResponse mutateInstance(IndexLifecycleExplainResp
                 return IndexLifecycleExplainResponse.newUnmanagedIndexResponse(index);
+            case 11:
+                isTransitiveError = true;
+                failedStepRetryCount = randomInt(13);
```
This could randomly cause a test failure if the count were the same, it should use

```suggestion
                failedStepRetryCount = randomValueOtherThan(failedStepRetryCount, () -> randomInt(13));
```

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/OpenFollowerIndexStepTests.java:
```diff
@@ -51,10 +51,35 @@ protected OpenFollowerIndexStep copyInstance(OpenFollowerIndexStep instance) {
         return new OpenFollowerIndexStep(instance.getKey(), instance.getNextStepKey(), instance.getClient());
     }
 
+    public void testOpenFollowerIndexIsNoopForAlreadyOpenIndex() {
```
Same comment above about this being in a separate PR

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -154,6 +156,48 @@ public void onFailure(Exception e) {
+        }
+
+        if (failedStep.isRetryable()) {
+            if (lifecycleState.isTransitiveError()) {
```
This hopefully would never be `null`, but maybe we should check for it?

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -154,6 +156,48 @@ public void onFailure(Exception e) {
+            if (lifecycleState.isTransitiveError()) {
+                int retryCount = lifecycleState.getFailedStepRetryCount() == null ? 0 : lifecycleState.getFailedStepRetryCount();
+                Integer maxRetriesCount =
+                    indexMetaData.getSettings().getAsInt(LifecycleSettings.LIFECYCLE_MAX_FAILED_STEP_RETRIES_COUNT, 15);
```
This shouldn't specify a default since a default is already specified in the setting itself (otherwise we have to change two places if it's changed in the future)

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -219,7 +263,7 @@ public void runPolicyAfterStateChange(String policy, IndexMetaData indexMetaData
             return;
         } else if (currentStep instanceof ErrorStep) {
-            logger.debug("policy [{}] for index [{}] on an error step, skipping execution", policy, index);
+            onErrorMaybeRetryFailedStep(policy, indexMetaData);
```
I actually think we should go back to just skipping it here. We can leave automatic ILM retries happening *only* on the periodic ILM interval (every 10 minutes by default) rather than every time state changes. This will work for our "backoff" of automatic retries.

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -290,6 +334,20 @@ static Step getCurrentStep(PolicyStepsRegistry stepRegistry, String policy, Inde
+            nextStepKey, nowSupplier, forcePhaseDefinitionRefresh);
+    }
+
+    private static void validateTransition(String indexName, ClusterState currentState, StepKey currentStepKey, StepKey nextStepKey,
```
Do we have tests for this? Would you mind making this package-visible and adding a unit test for its validation? (Can be done in a separate PR)

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -340,34 +393,65 @@ static ClusterState moveClusterStateToErrorStep(Index index, ClusterState cluste
+        return moveClusterStateToFailedStep(currentState, index, true);
+    }
+
+    private ClusterState moveClusterStateToFailedStep(ClusterState currentState, String index, boolean isRetry) {
```
I think it's getting confusing having `moveClusterStateToErrorStep` and `moveClusterStateToFailedStep`. The names are very similar and it's difficult to tell at a glance how the two differ.

How about we rename this to something like `moveClusterStateOutOfErrorStep` so it makes it clearer that one is entering the error state and the other is exiting the error state?

dakrone replied:
Also, the `isRetry` is an overloaded term since we have a "retry" API, maybe it should be `isAutomaticRetry`?

andreidan replied:
Thinking about this rename some more, it seems to me that `moveClusterStateToFailedStep` is actually more descriptive than `moveClusterStateOutOfErrorStep` (seeing a method like this, I'd immediately ask myself "moving where?" and I'd have to look at what it does to answer that question)

I appreciate the number of `moveClusterStateToX` methods increased with this PR. The methods we currently have are `moveClusterStateTo`: 
* `FailedStep`
* `ErrorStep`
* `NextStep`
* `Step`
* `RetryFailedStep` - added in this PR but it's just an overload which I'll drop as there's no need for an extra term in the vocabulary

Unfortunately, I don't have a better suggestion to clean up the namings (probably because I've spent a lot of time recently in this code and it makes sense to me) so if `moveClusterStateOutOfErrorStep` makes more sense to you I'll rename it, just wanted to express my concerns before doing so. @dakrone @gwbrown

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunnerTests.java:
```diff
@@ -998,6 +1120,51 @@ public void testMoveClusterStateToFailedNotOnError() {
+        assertThat(executionState.getFailedStepRetryCount(), is(1));
+    }
+
+    private static final class RetryableMockStep extends MockStep {
```
Minor nit, but can you move this to the end of the file so it's not interspersed with the other `test*` tests?

dakrone reviewed (changes requested):
Thanks for working on this Andrei, I have some comments about it.

With regard to the number of retries, I think if we drop the retries to be on the periodic interval (so every 10 minutes by default), we should change the setting to -1 as a default, meaning to try forever. This lets us keep the ability for a user to set retries not to run infinitely, but still have the default keep trying forever by default (albeit spaced out by 10 minutes between each retry).

AthenaEryma reviewed (commented):
Left some comments - I tried to deduplicate my comments that overlapped with Lee's, but may have missed some.

Another couple thoughts:
1) I'm not sure `transitive` is the right adjective here as it doesn't clearly communicate what the intent is in the code.  I think `autoretryable` or something may be a better choice.
2) While I'm with you on infinite retries being a little scary, it does limit the usefulness of auto-retry in some cases. For example, if an index has a misconfigured alias (a common issue in the field), it would be nice if ILM could figure itself out automatically once the configuration issue is resolved. But the automatic retries will run out in 2.5 hours, which is a fairly short time window to allow for an administrator to notice and correct the issue.  IMO it's worth thinking about if we need to cap the retries, and even if we do in general, if some errors can ignore the cap.

andreidan commented:
@gwbrown @dakrone thanks for the review. I'll move on to implement your suggestions. I'll create a separate PR for the open/close follower steps changes (made them in order to make the steps a bit more resilient as the api calls would fail when executed against a read-only index for example and we execute them regardless if the index is already in the desired state ie. open or closed)

andreidan committed:
Add versioning protection.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retry-failed-step

andreidan committed:
Guard the serialisation changes against gte 8.0.0

andreidan commented:
@elasticmachine retest this please

andreidan commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -154,6 +156,48 @@ public void onFailure(Exception e) {
+                        public void onFailure(String source, Exception e) {
+                            logger.error("retry execution of step [{}] failed due to [{}]", failedStep.getKey().getName(), e);
+                        }
+                    });
```
We'll follow up on running async steps ( https://github.com/elastic/elasticsearch/pull/48010#issuecomment-541796585 )

andreidan committed:
Rename isTransitiveError to isAutoRetryableError

andreidan committed:
Fix the ILM explain tet

andreidan committed:
IndexLifecycleExplainResponseTest: adjust the random retry values

andreidan commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponseTests.java:
```diff
@@ -52,6 +52,8 @@ private static IndexLifecycleExplainResponse randomManagedIndexExplainResponse()
             stepNull ? null : randomAlphaOfLength(10),
             randomBoolean() ? null : randomAlphaOfLength(10),
+            stepNull ? null : randomBoolean(),
+            stepNull ? null : randomInt(15),
```
this is somewhat arbitrary (I set it up to the previous rather arbitrary retry count I configured). I'll change it to 10 to not stand out of the bunch (especially as the retry count default will be infinite)

andreidan committed:
Dorp lifecycle poll interval configuration in IT

andreidan committed:
Change max retry setting default to -1

andreidan committed:
Log the index name too

andreidan committed:
On state change don't attempt to retry failed step

andreidan committed:
Drop moveClusterStateToRetryFailedStep as it was just an overload

andreidan committed:
Test validateTransition

andreidan committed:
Revert "ILM open/close steps are noop if idx is open/close"

This reverts commit 3fb77aaf4e6fc33f2091fde5d41f0e1955928804.

andreidan committed:
Update isRetryable javadoc

Co-Authored-By: Lee Hinman <dakrone@users.noreply.github.com>

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retry-failed-step

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/MoveToErrorStepUpdateTaskTests.java:
```diff
@@ -43,6 +44,7 @@
     String policy;
     ClusterState clusterState;
     Index index;
+    private PolicyStepsRegistry stepRegistry;
```
I don't think this is used anywhere?

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunnerTests.java:
```diff
@@ -1227,6 +1386,74 @@ public void testIsReadyToTransition() {
+        try {
+            IndexLifecycleRunner.validateTransition(indexMetaData, currentStepKey, nextStepKey, policyRegistry);
+        } catch (Exception e) {
+            fail("validateTransition should not throw exception on valid transitions");
```
Can you add `logger.error(e);` above this so we don't accidentally swallow the stacktrace for debugging if it fails in the future?

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -154,6 +156,50 @@ public void onFailure(Exception e) {
+        }
+
+        if (failedStep.isRetryable()) {
+            if (lifecycleState.isAutoRetryableError() != null && lifecycleState.isAutoRetryableError()) {
```
The more I thing about these the more I think we should stick with only checking one here. I think only checking the lifecycleState one may be the way we want to go.

dakrone reviewed (changes requested):
I left some more feedback on this, I think it's looking good, just a few things to take care of now.

jasontedor commented:
> Controlling the number of retries is mostly for this particular use case where we'll allow 1-2 retries in tests before the `error` state is reached. I don't think there's any production use case where a user would want to modify this setting.

I don’t think this should be a user-settable setting. It’s possible to dictate this by a setting that is only registered in tests. Can we take that approach here and avoid making this user-configurable, and all the discussion around it.

andreidan committed:
Remove the LIFECYCLE_MAX_FAILED_STEP_RETRIES_COUNT setting.

andreidan committed:
Drop unused field

andreidan committed:
Guard against a possible null failed step

andreidan committed:
Throw IndexNotFoundException instead of IllegalArgumentException

andreidan committed:
Rename to moveClusterStateToPreviouslyFailedStep

andreidan committed:
Don't use randomBool as we're asserting only the managed case

andreidan committed:
Test the ilm/explain retry count output in separate test

andreidan committed:
Fix test to expect IndexNotFoundException

andreidan committed:
Update retry api test to expect index not found exception

andreidan commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:
```diff
@@ -408,7 +409,7 @@ ClusterState moveClusterStateToFailedStep(ClusterState currentState, String inde
         IndexMetaData indexMetaData = currentState.metaData().index(index);
         if (indexMetaData == null) {
-            throw new IllegalArgumentException("index [" + index + "] does not exist");
+            throw new IndexNotFoundException(index);
```
This has the implication that some APIs (ie. `ilm/_retry`) will return `404` instead of `400` which is a breaking change. I don't think we should make this change as part of this PR. Thoughts @dakrone ?

dakrone replied:
We can stick with the existing exception, we can discuss making this change as a separate PR at a later time, thanks for catching this.

andreidan replied:
Pushed the reverts for the exception change

andreidan commented:
Thanks for the review. I think I addressed all your comments and besides a question on an exception change that has impact on the rest apis I think this is ready.

andreidan committed:
Revert "Update retry api test to expect index not found exception"

This reverts commit 6a4e3ebba443762632178d39b207b2d5f689ec4a.

andreidan committed:
Revert "Fix test to expect IndexNotFoundException"

This reverts commit 998eaa980005d9dcac2ec4ee365cd33339519df8.

andreidan committed:
Revert "Throw IndexNotFoundException instead of IllegalArgumentException"

This reverts commit 90cfb011512a79eafac6b4cf05dbbe09604d0dee.

dakrone reviewed (approved):
LGTM, thanks for iterating on this @andreidan!

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retry-failed-step
